### PR TITLE
Delete core instance

### PR DESCRIPTION
**Pull request by gabrielbussolo**
# Summary of this proposal

## Issue(s) number(s)

*Disclaimer: Please do not create a Pull Request without creating an issue first.*

## Checklist for submitter

- [ ] My PR has a related issue/bug number.
- [ ] My PR provides tests
  - [ ] Integration tests are added/passes
  - [ ] Unit tests are added/passes
  - [ ] End-to-end tests added/passes
  - [ ] Static code check added/passes
  - [ ] Linting on documentation added/passes
  - [ ] Does not affect code coverage stats
- [ ] My PR requires updating dependencies
- [ ] My PR has the documentation changes required.

## Checklist for reviewer

- [ ] The proposal fixes a bug/issue or implements a new feature that is well described.
- [ ] The proposal has sufficient test cases that covers the changes.
  - [ ] If changes an API, it does not break backwards compatibility (-1 major version).
- [ ] If integration is required, the proposal has integration tests.
- [ ] The proposal does not break coverage stats
- [ ] The proposal has the required documentation changes.

## Backport

- [ ] Backport to the latest stable release.

### CODE REVIEW

**Comment by niedbalski** [comment]
@gabrielbussolo is this the branch you referred in the sync today?

**Comment by gabrielbussolo** [comment]
> @gabrielbussolo is this the branch you referred in the sync today?

yes, sorry i didnt updated yesterday.

still need improvementes here.
im working to use same schema that u used to create a k8sclient :)

**Comment by gabrielbussolo** [comment]
I will work to improve the delete methods, and delete the health-check and loadbalancer instances

File: cmd/calyptia/delete_core_instance.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+			}
+
+			label := "aggregatorID=" + aggregator.ID
+
+			deployments, err := k8sClient.FindDeploymentsByLabel(ctx, label)
+			if err != nil {
+				return err
+			}
+
+			for _, item := range deployments.Items {
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:74]: Isn't it possible to directly call delete without listing them first 🤔

Comment (nicolasparada) [reply cmd/calyptia/delete_core_instance.go:74]: There is a `DeleteCollection` method that should do it.

File: cmd/calyptia/delete_core_instance.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	cmd := &cobra.Command{
+		Use:   "instance",
+		Short: "Delete a core instance",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := context.Background()
+			if aggregatorName == "" {
+				return fmt.Errorf("the core instance name must be provided")
+			}
+			agg, err := config.cloud.Aggregators(ctx, config.projectID, types.AggregatorsParams{
+				Name: &aggregatorName,
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:28]: On other commands we can find by either ID or name.
That a look at how is done in other commands.

The arg can be autocompleted too.

File: cmd/calyptia/delete_core_instance.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func newCmdDeleteCoreInstance(config *config, testClientSet kubernetes.Interface) *cobra.Command {
+	var aggregatorName string
+	var skipError bool
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	configOverrides := &clientcmd.ConfigOverrides{}
+	cmd := &cobra.Command{
+		Use:   "instance",
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:20]: The aggregator should come from the arguments and not from a flag.
See other delete commands.

File: cmd/calyptia/delete_core_instance.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	"github.com/calyptia/api/types"
+	"github.com/calyptia/cli/k8s"
+	"github.com/spf13/cobra"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func newCmdDeleteCoreInstance(config *config, testClientSet kubernetes.Interface) *cobra.Command {
+	var aggregatorName string
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:15]:
```suggestion
```

Not needed if you take it from the args.

File: cmd/calyptia/delete_core_instance.go (hunk 2 above)
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:28]:
```suggestion
```

File: cmd/calyptia/delete_core_instance.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+		Use:   "instance",
+		Short: "Delete a core instance",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := context.Background()
+			if aggregatorName == "" {
+				return fmt.Errorf("the core instance name must be provided")
+			}
+			agg, err := config.cloud.Aggregators(ctx, config.projectID, types.AggregatorsParams{
+				Name: &aggregatorName,
+			})
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:29]:
```suggestion
```

File: cmd/calyptia/delete_core_instance.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+			err = config.cloud.DeleteAggregator(ctx, aggregator.ID)
+			if err != nil {
+				return err
+			}
+			fmt.Fprintf(cmd.OutOrStdout(), "%q was deleted successfully on cloud\n", aggregatorName)
+			return nil
+		},
+	}
+	fs := cmd.Flags()
+	fs.StringVar(&aggregatorName, "name", "", "Core instance name (autogenerated if empty)")
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:152]:
```suggestion
```
Taken from args.

File: cmd/calyptia/delete_core_instance.go (hunk 1 above)
Comment (gabrielbussolo) [reply cmd/calyptia/delete_core_instance.go:74]: i couldnt find a way to delete by "tag" so im listing to get the health-check name, so bcz i have too do it for the health-check, im getting the deployment also

**Comment by niedbalski** [comment]
I am awaiting on @nicolasparada's comments to be addressed before jumping to review @gabrielbussolo

File: cmd/calyptia/delete_core_instance.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+		Short:             "Delete a core instance",
+		Args:              cobra.ExactArgs(1),
+		ValidArgsFunction: config.completeAggregators,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := context.Background()
+
+			aggregatorKey := args[0]
+			var agg types.Aggregator
+			var err error
+			if validUUID(aggregatorKey) {
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:30]:
```suggestion
```
A name could have UUID format.
That's why `loadAggregatorID` accepts just a "key" that could be the name or ID and does the resolution and gives you back the ID.

File: cmd/calyptia/delete_core_instance.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+				agg, err = config.cloud.Aggregator(ctx, aggregatorKey)
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:31]:
```suggestion
```

File: cmd/calyptia/delete_core_instance.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+				if err != nil {
+					return err
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:33]:
```suggestion
```

File: cmd/calyptia/delete_core_instance.go [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+				}
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:34]:
```suggestion
```

File: cmd/calyptia/delete_core_instance.go [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+			} else {
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:35]:
```suggestion
```

File: cmd/calyptia/delete_core_instance.go [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+				aggregatorID, err := config.loadAggregatorID(aggregatorKey)
+				if err != nil {
+					return err
+				}
+				agg, err = config.cloud.Aggregator(ctx, aggregatorID)
+				if err != nil {
+					return err
+				}
+			}
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:44]:
```suggestion
```

File: cmd/calyptia/delete_core_instance.go [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+				if err != nil {
+					if !skipError {
+						return err
+					}
+					fmt.Fprintf(cmd.OutOrStdout(), "%q was not deleted, err: %v\n", item.Name, err)
+				}
+				fmt.Fprintf(cmd.OutOrStdout(), "service=%q deleted\n", item.Name)
+			}
+
+			clusterRole := agg.Name + "-cluster-role"
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:112]: Shouldn't all resources be deleted by labels?

**Comment by gabrielbussolo** [comment]
> Shouldn't all resources be deleted by labels?

i noticed that we can delete deployments by label, we r deleting the core instance and the health-check, so if some error occur i cant know which one failed to tell the user on `--skip-error`

some resources like services, dont have "deleteCollection" option that i can pass the tag, thats why i standardized all by name to dont have some things being deleted by tag, others by name.

what u guys think? @nicolasparada @niedbalski

File: cmd/calyptia/delete_core_instance.go (hunk 13 above)
Comment (gabrielbussolo) [reply cmd/calyptia/delete_core_instance.go:97]: i  changed all for delete by label, but maintained the code for logging :)

**Comment by gabrielbussolo** [comment]
this current implementation accumplish the goal to delete by label.
But i noticed a behavior that idk if we want it, deleting by label it doest return an error if the resource doest not exist. So doest make sense log each resource that is being deleted once that we cant know if it got deleted or not.

ex: when deleting by name the api returns an error if thje resource doest not exist, deleting by label not.

knowing this `--skip-error` stop make sense i guess once delete by collection will return error in few specific errors

@niedbalski @nicolasparada

**Review by niedbalski — CHANGES_REQUESTED** [review]
I hit an issue while running the delete on deployments that have been made in a different namespace (-n test), https://calyptia.slack.com/archives/C02SML63BQA/p1656576134963649 , we shouldn't scope deletion of deployments and other objects to the current client provided namespace as it can be different to the one on which we deployed.

It's entirely safe to assume that the labels are unique across the different cluster namespaces, so the proposal (if not possible to apply a delete on deployments that isn't namespace scoped) is to iterate across the namespaces available on the cluster and find all the objects that have the calyptia_*labels with the specified id(s).

Also, this PR is missing the --yes confirmation, is not provided it should ask for y/n not going straight on deletion.

File: k8s/labels.go [hunk 14]
```diff
@@ -10,7 +10,7 @@ const (
 
 const (
 	LabelProjectID      = "calyptia_project_id"
-	LabelAggregatorID   = "calyptia_aggregator_id"
+	LabelAggregatorID   = "aggregatorID"
```
Comment (niedbalski) [inline k8s/labels.go:13]: revert this to the original value, please.

File: cmd/calyptia/delete_core_instance.go [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+					return err
+				}
+				fmt.Fprintf(cmd.OutOrStdout(), "a problem occurred while deleting %s, err: %v\n", roleBinding, err)
+			}
+
+			err = config.cloud.DeleteAggregator(ctx, agg.ID)
+			if err != nil {
+				return err
+			}
+			fmt.Fprintf(cmd.OutOrStdout(), "%q was deleted successfully on cloud\n", agg.Name)
```
Comment (niedbalski) [inline cmd/calyptia/delete_core_instance.go:154]: calyptia-core instance %q was successfully deleted.

File: cmd/calyptia/delete_core_instance.go [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+	fs := cmd.Flags()
+	fs.BoolVar(&skipError, "skip-error", false, "Skip errors during delete process")
+	//isNonInteractiveMode := os.Stdin == nil || !term.IsTerminal(int(os.Stdin.Fd()))
+	//fs.BoolVar(&confirmDelete, "yes", isNonInteractiveMode, "Confirm deletion")
+	clientcmd.BindOverrideFlags(configOverrides, fs, clientcmd.RecommendedConfigOverrideFlags("kube-"))
+	return cmd
+}
+
+func listDeletionsByLabel(ctx context.Context, k8sClient *k8s.Client, cmd *cobra.Command, label string) error {
+	namespaces, err := k8sClient.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
```
Comment (nicolasparada) [inline cmd/calyptia/delete_core_instance.go:169]: So, the user sets the namespace with the flag `--kube-namespace` but here it is using all namespaces, so effectively ignoring it.

Comment (niedbalski) [reply cmd/calyptia/delete_core_instance.go]: It doesn't matter, we'll iterate across all namespaces, including what is in kube-namespace provided by the user.

**Comment by nicolasparada** [comment]
All right, looks good to me.
There is just a merge conflict.

**Comment by niedbalski** [comment]
@gabrielbussolo can you fix the conflicts?

**Comment by niedbalski** [comment]
LGTM, I added a couple of format commits and fixed the merge conflict.

**Comment by niedbalski** [comment]
@gabrielbussolo @nicolasparada please take a look at #75 this afternoon.

**Review by niedbalski — APPROVED** [review]

**Pull request merged**